Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/Allocator/Jobs/ScheduledRides/ScheduledRideNotificationsToDriver.hs

```diff
@@ -0,0 +1,111 @@
+
+ or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more details. You should have received a copy of
+
+ the GNU Affero General Public License along with this program. If not, see <https://www.gnu.org/licenses/>.
+-}
+
+module SharedLogic.Allocator.Jobs.ScheduledRides.ScheduledRideNotificationsToDriver where
+
+import qualified Data.Aeson as A
+import qualified Data.Map as M
+import qualified Data.Text as T
+import qualified Domain.Types.Booking as DB
+import qualified Domain.Types.Common as DTC
+import qualified Domain.Types.DriverInformation as DInfo
+import qualified Domain.Types.Merchant.Overlay as DOverlay
+import Domain.Types.RideRelatedNotificationConfig
+import qualified Kernel.Beam.Functions as B
+import Kernel.External.Encryption (decrypt)
+import Kernel.External.Types (Language (..))
+import Kernel.Prelude
+import Kernel.Sms.Config (SmsConfig)
+import Kernel.Types.Error
+import Kernel.Types.SlidingWindowLimiter (APIRateLimitOptions)
+import Kernel.Utils.Common
+import Lib.Scheduler
+import SharedLogic.Allocator
+import qualified Storage.CachedQueries.Merchant.MerchantMessage as CMM
+import qualified Storage.CachedQueries.Merchant.MerchantPushNotification as CPN
+import qualified Storage.CachedQueries.Merchant.Overlay as CMO
+import qualified Storage.Queries.Booking as QB
+import qualified Storage.Queries.DriverInformation.Internal as QDInfo
+import qualified Storage.Queries.Person as QPerson
+import Tools.Error
+import Tools.Notifications
+import qualified Tools.SMS as Sms
+
+sendScheduledRideNotificationsToDriver ::
+  ( EncFlow m r,
+    CacheFlow m r,
+    MonadFlow m,
+    EsqDBFlow m r,
+    HasFlowEnv m r ["apiRateLimitOptions" ::: APIRateLimitOptions, "smsCfg" ::: SmsConfig],
+    HasField "schedulerSetName" r Text,
+    HasField "schedulerType" r SchedulerType,
+    HasField "jobInfoMap" r (M.Map Text Bool)
+  ) =>
+  Job 'ScheduledRideNotificationsToDriver ->
+  m ()
+sendScheduledRideNotificationsToDriver Job {id, jobInfo} = withLogTag ("JobId-" <> id.getId) do
+  let jobData = jobInfo.jobData
+      merchantOpCityId = jobData.merchantOperatingCityId
+      bookingId = jobData.bookingId
+      driverId = jobData.driverId
+      notificationType = jobData.notificationType
+      notificationKey = jobData.notificationKey
+      onlyIfOffline = jobData.onlyIfOffline
+  smsCfg <- asks (.smsCfg)
+  booking <- QB.findById bookingId >>= fromMaybeM (BookingDoesNotExist bookingId.getId)
+  driverInfo <- B.runInReplica (QDInfo.getDriverInfo driverId.getId) >>= fromMaybeM DriverInfoNotFound
+  let isNotificationRequired = not onlyIfOffline || (driverInfo.mode /= Just DInfo.ONLINE)
+
+  when (isNotificationRequired && booking.status == DB.TRIP_ASSIGNED) do
+    driver <- B.runInReplica $ QPerson.findById driverId >>= fromMaybeM (PersonDoesNotExist driverId.getId)
+    mobileNumber <- mapM decrypt driver.mobileNumber >>= fromMaybeM (PersonFieldNotPresent "mobileNumber")
```

Comment:
Fetch only in case needed.

Reply:
it will be required for CALL notification also